Sources_v3/StreamChat/ChatClient_Tests.swift

```diff
@@ -373,58 +374,345 @@ class ChatClient_Tests: StressTestCase {
+                // Whenever `tokenProvider` is called we generate new token for the current user.
+                let updatedToken = Token.unique(userId: currentUserId)
+                lastProvidedToken = updatedToken
+                $1(.success(updatedToken))
+            },
             workerBuilders: [TestWorker.init],
             eventWorkerBuilders: [TestEventWorker.init],
-            environment: testEnv.environment
+            environment: .mock,
+            completion: { error in
+                XCTAssertNil(error)
+                completionCalled = true
+            }
         )
-        
-        let currentUserController = client.currentUserController()
 
-        // Generate userIds for first user
-        let oldUserId: UserId = .unique
-        let oldUserToken: Token = .unique
-        
-        // Set a user
-        currentUserController.setUser(userId: oldUserId, name: nil, imageURL: nil, token: oldUserToken)
+        // Simulate web-socket connection established event.
+        let connectionId: String = .unique
+        client.webSocketClient?.simulateConnectionStatus(.connected(connectionId: connectionId))
+        // Wait the client to be initialized.
+        AssertAsync.willBeTrue(completionCalled)
+
+        // Assert token is set.
+        XCTAssertEqual(client.currentToken, lastProvidedToken)
+        // Assert current user is set.
+        XCTAssertEqual(client.currentUserId, currentUserId)
+        // Assert connection is established.
+        XCTAssertEqual(client.connectionId, connectionId)
+        XCTAssertEqual(client.connectionStatus, .connected)
+        // Assert background workers are created.
+        XCTAssertEqual(client.backgroundWorkers.count, 2)
 
         // Save worker's UUID
         let oldWorkerUUID = (client.backgroundWorkers.first as! TestWorker).id
         let oldEventWorkerUUID = (client.backgroundWorkers.last as! TestEventWorker).id
-        
-        // Set the same user again
-        currentUserController.setUser(userId: oldUserId, name: nil, imageURL: nil, token: oldUserToken)
 
-        // .. to make sure worker's are not re-created for the same user
+        // Reset disconnect counter.
+        client.mockWebSocketClient.disconnect_calledCounter = 0
+
+        // Call `reloadUserIfNeeded` and catch the error.
+        var reloadUserIfNeededCompletionCalled = false
+        client.reloadUserIfNeeded { error in
+            XCTAssertNil(error)
+            reloadUserIfNeededCompletionCalled = true
+        }
+
+        // Wait `reloadUserIfNeeded` to be completed.
+        AssertAsync.willBeTrue(reloadUserIfNeededCompletionCalled)
+        // Assert token is updated.
+        XCTAssertEqual(client.currentToken, lastProvidedToken)
+        // Assert current user is not changed.
+        XCTAssertEqual(client.currentUserId, currentUserId)
+        // Assert background workers are not recreated.
+        XCTAssertEqual(client.backgroundWorkers.count, 2)
```

Comment:
How do you know the workers weren't recreated? What if these are two different workers? :)